File: hammer-gui/src/main/java/dev/vini2003/hammer/gui/api/common/widget/Widget.java
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.gui.api.common.widget;

import com.google.common.collect.ImmutableList;
import dev.vini2003.hammer.core.HC;
import dev.vini2003.hammer.core.api.client.util.PositionUtil;
import dev.vini2003.hammer.core.api.common.math.position.Position;
import dev.vini2003.hammer.core.api.common.math.position.StaticPosition;
import dev.vini2003.hammer.core.api.common.math.position.Positioned;
import dev.vini2003.hammer.core.api.common.math.size.Size;
import dev.vini2003.hammer.core.api.common.math.size.Sized;
import dev.vini2003.hammer.core.api.common.tick.Ticks;
import dev.vini2003.hammer.gui.api.common.event.*;
import dev.vini2003.hammer.gui.api.common.event.base.Event;
import dev.vini2003.hammer.gui.api.common.event.type.EventType;
import dev.vini2003.hammer.gui.api.common.listener.EventListener;
import dev.vini2003.hammer.gui.registry.common.HGUINetworking;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.OrderedText;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.function.Supplier;

public abstract class Widget implements Positioned, Sized, EventListener, Ticks {
	protected Position position = Position.of(0.0F, 0.0F);
	protected Size size = Size.of(0.0F, 0.0F);
	
	protected WidgetCollection collection;
	protected WidgetCollection.Root rootCollection;
	
	protected Widget parent;
	
	protected Supplier<List<OrderedText>> tooltip = ImmutableList::of;
	
	protected boolean hidden = false;
	protected boolean focused = false;
	protected boolean held = false;
	protected boolean locking = false;
	
	protected Map<EventType, Collection<EventListener<?>>> listeners = new HashMap<>();
	
	/**
	 * The widget's ID is used to identify it when
	 * synchronizing events between the client and the server.
	 */
	@Nullable
	protected String id;
	
	protected Widget() {
		onEvent(EventType.ADDED, this::onAdded);
		onEvent(EventType.REMOVED, this::onRemoved);
		onEvent(EventType.FOCUS_GAINED, this::onFocusGained);
		onEvent(EventType.FOCUS_RELEASED, this::onFocusReleased);
		onEvent(EventType.KEY_PRESSED, this::onKeyPressed);
		onEvent(EventType.KEY_RELEASED, this::onKeyReleased);
		onEvent(EventType.CHARACTER_TYPED, this::onCharacterTyped);
		onEvent(EventType.MOUSE_CLICKED, this::onMouseClicked);
		onEvent(EventType.MOUSE_DRAGGED, this::onMouseDragged);
		onEvent(EventType.MOUSE_MOVED, this::onMouseMoved);
		onEvent(EventType.MOUSE_RELEASED, this::onMouseReleased);
		onEvent(EventType.MOUSE_SCROLLED, this::onMouseScrolled);
		onEvent(EventType.LAYOUT_CHANGED, this::onLayoutChanged);
	}
	
	protected void onAdded(AddedEvent event) {
		rootCollection = event.rootCollection();
		collection = event.collection();
	}
	
	protected void onRemoved(RemovedEvent event) {
		rootCollection = null;
		collection = null;
	}
	
	protected void onFocusGained(FocusGainedEvent event) {
	
	}
	
	protected void onFocusReleased(FocusReleasedEvent event) {
	
	}
	
	protected void onKeyPressed(KeyPressedEvent event) {
	
	}
	
	protected void onKeyReleased(KeyReleasedEvent event) {
	
	}
	
	protected void onCharacterTyped(CharacterTypedEvent event) {
	
	}
	
	protected void onMouseClicked(MouseClickedEvent event) {
	
	}
	
	protected void onMouseDragged(MouseDraggedEvent event) {
	
	}
	
	protected void onMouseMoved(MouseMovedEvent event) {
		updateFocus(event.x(), event.y());
	}
	
	protected void onMouseReleased(MouseReleasedEvent event) {
	
	}
	
	protected void onMouseScrolled(MouseScrolledEvent event) {
	
	}
	
	protected void onLayoutChanged(LayoutChangedEvent event) {
		if (rootCollection != null && rootCollection.isClient()) {
			updateFocus(PositionUtil.getMouseX(), PositionUtil.getMouseY());
		}
	}
	
	/**
	 * <p>Returns whether the given event type should be synced.</p>
	 *
	 * @param type the type.
	 *
	 * @return the valid.
	 */
	public boolean shouldSync(EventType type) {
		return false;
	}
	
	/**
	 * <p>Adds a listener for a given event type to this widget.</p>
	 *
	 * @param type     the type.
	 * @param listener the listener.
	 */
	public <T extends Event> void onEvent(EventType type, EventListener<T> listener) {
		listeners.computeIfAbsent(type, ($) -> new ArrayList<>());
		listeners.get(type).add(listener);
	}
	
	/**
	 * <p>Returns all listeners of for given event type in this widget.</p>
	 *
	 * @param type the type.
	 *
	 * @return the listeners.
	 */
	public <T extends Event> Collection<EventListener<T>> getListeners(EventType type) {
		listeners.computeIfAbsent(type, ($) -> new ArrayList<>());
		
		return (Collection) listeners.get(type);
	}
	
	/**
	 * <p>Called when an event is dispatched.</p>
	 *
	 * <p>Syncs it if this widget is attached to a screen or screen handler,
	 * and is running on the client.</p>
	 *
	 * @param event the event.
	 */
	@Override
	public void dispatchEvent(Event event) {
		if (!allowsEvents()) {
			return;
		}
		
		var rootCollection = getRootCollection();
		
		// Always propagate Layout Changed, as some widgets may
		// handle special logic in there that will not run on the server
		// if it is not propagated correctly.
		if ((shouldSync(event.type()) || event.type() == EventType.LAYOUT_CHANGED) && (isFocused() || rootCollection != null && rootCollection.isScreenHandler() && rootCollection.isClient())) {
			if (id == null) {
				HC.LOGGER.debug("A widget was not given an ID, but tried to synchronize an event!");
				
				return;
			}
			
			var buf = PacketByteBufs.create();
			buf.writeString(id);
			
			event.writeToBuf(buf);
			
			ClientPlayNetworking.send(HGUINetworking.SYNC_WIDGET_EVENT, buf);
		}
		
		for (var listener : getListeners(event.type())) {
			listener.dispatchEvent(event);
		}
		
		if (this instanceof WidgetCollection collection && (rootCollection == null || rootCollection.isClient())) {
			if (event.type() != EventType.ADDED && event.type() != EventType.REMOVED) {
				for (var child : collection.getChildren()) {
					child.dispatchEvent(event);
				}
			}
		}
	}
	
	/**
	 * Updates whether the widget is focused or not.
	 */
	protected void updateFocus(float x, float y) {
		var wasFocused = isFocused();
		var isFocused = isPointWithin(x, y);
		
		var isBelowOtherWidget = false;
		
		for (var child : getRootCollection().getAllChildren()) {
			if (child == this) {
				continue;
			}
			
			if (child.isPointWithin(x, y) && child.getZ() > getZ()) {
				isBelowOtherWidget = true;
				
				break;
			}
		}
		
		if (isBelowOtherWidget) {
			isFocused = false;
		}
		
		setFocused(isFocused);
		
		if (wasFocused && !isFocused) {
			dispatchEvent(new FocusReleasedEvent());
		}
		
		if (!wasFocused && isFocused) {
			dispatchEvent(new FocusGainedEvent());
		}
	}
	
	/**
	 * Returns whether a point is within this widget's boundaries or not.
	 *
	 * @param x the point's X component.
	 * @param y the point's Y component.
	 *
	 * @return the valid.
	 */
	public boolean isPointWithin(float x, float y) {
		return x > getX() && x < getX() + getWidth() && y > getY() && y < getY() + getHeight();
	}
	
	/**
	 * Configures transforms before drawing.
	 * Useful for translating to the correct Z layer.
	 *
	 * @param context	the draw context.
	 * @param tickDelta the tick delta.
	 */
	@Environment(EnvType.CLIENT)
	public void onBeginDraw(DrawContext context, float tickDelta) {
		context.getMatrices().push();
		context.getMatrices().translate(0.0F, 0.0F, getZ());
	}
	
	/**
	 * Reverts transforms after drawing.
	 *
	 * @param context	the draw context.
	 * @param tickDelta the tick delta.
	 */
	@Environment(EnvType.CLIENT)
	public void onEndDraw(DrawContext context, float tickDelta) {
		context.getMatrices().pop();
	}
	
	/**
	 * Draws this widget.
	 *
	 * @param context	the draw context.
	 * @param tickDelta	the tick delta.
	 */
	@Environment(EnvType.CLIENT)
	public void draw(DrawContext context, float tickDelta) {
		onBeginDraw(context, tickDelta);
		
		if (this instanceof WidgetCollection collection) {
			for (var child : collection.getChildren()) {
				if (!child.isHidden()) {
					child.draw(context, tickDelta);
				}
			}
		}
		
		onEndDraw(context, tickDelta);
	}
	
	@Deprecated
	public List<Text> getTooltips() {
		return Collections.emptyList();
	}
	
	/**
	 * Returns this widget's tooltips.
	 *
	 * @return the tooltips.
	 */
	public List<OrderedText> getTooltip() {
		if (tooltip != null) {
			return tooltip.get();
		}
		
		return new ArrayList<>();
	}
	
	@Override
	public Position getPosition() {
		return position;
	}
	
	@Override
	public void setPosition(Position position) {
		this.position = position;
		
		dispatchEvent(new LayoutChangedEvent());
	}
	
	@Override
	public Size getSize() {
		if (size.getWidth() == 0.0F && size.getHeight() == 0.0F) {
			size = getStandardSize();
		}
		
		return size;
	}
	
	@Override
	public void setSize(Size size) {
		this.size = size;
		
		dispatchEvent(new LayoutChangedEvent());
	}
	
	@Nullable
	public String getId() {
		return this.id;
	}
	
	public void setId(@Nullable String id) {
		this.id = id;
	}
	
	public boolean allowsEvents() {
		return !(this instanceof WidgetCollection.Root) || collection != null || rootCollection != null;
	}
	
	public WidgetCollection getCollection() {
		return collection;
	}
	
	public void setCollection(WidgetCollection collection) {
		this.collection = collection;
	}
	
	public WidgetCollection.Root getRootCollection() {
		if (this instanceof WidgetCollection.Root rootCollection) {
			return rootCollection;
		}
		
		return rootCollection;
	}
	
	public void setRootCollection(WidgetCollection.Root rootCollection) {
		this.rootCollection = rootCollection;
	}
	
	public Widget getParent() {
		return parent;
	}
	
	public void setParent(Widget parent) {
		this.parent = parent;
	}
	
	public void setTooltip(Supplier<List<OrderedText>> tooltip) {
		this.tooltip = tooltip;
	}
	
	@Deprecated
	public void setTooltipSupplier(Supplier<List<Text>> tooltipSupplier) {
		if (tooltipSupplier.get() != null) {
			this.tooltip = () -> tooltipSupplier.get().stream().map(Text::asOrderedText).toList();
		}
	}
	
	public void setHidden(boolean hidden) {
		this.hidden = hidden;
	}
	
	public boolean isHidden() {
		if (parent != null) {
			return parent.isHidden() || hidden;
		} else {
			return hidden;
		}
	}
	
	public void setFocused(boolean focused) {
		this.focused = focused;
	}
	
	public boolean isFocused() {
		if (parent != null) {
			return !parent.isHidden() && !isHidden() && focused;
		} else {
			return !isHidden() && focused;
		}
	}
	
	public void setHeld(boolean held) {
		this.held = held;
	}
	
	public boolean isHeld() {
		if (parent != null) {
			return !parent.isHidden() && !isHidden() && held;
		} else {
			return !isHidden() && held;
		}
	}
	
	public void setLocking(boolean locking) {
		this.locking = locking;
	}
	
	public boolean isLocking() {
		return locking;
	}
}

File: src/main/kotlin/dev/vini2003/hammer/common/widget/Widget.kt
package dev.vini2003.hammer.common.widget

import dev.vini2003.hammer.common.geometry.position.Position
import dev.vini2003.hammer.common.geometry.position.Positioned
import dev.vini2003.hammer.common.geometry.size.Size
import dev.vini2003.hammer.common.geometry.size.Sized
import dev.vini2003.hammer.common.util.Networks
import dev.vini2003.hammer.common.util.Positions
import dev.vini2003.hammer.common.util.extension.plus
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import java.util.*
import kotlin.properties.Delegates

abstract class Widget : Positioned, Sized {
	override var position: Position by Delegates.observable(Position.of(0, 0)) { _, _, _ ->
		onLayoutChanged()
	}
	
	override var size: Size by Delegates.observable(Size.of(0, 0)) { _, _, _ ->
		onLayoutChanged()
	}

	var handled: WidgetCollection.Handled? = null
	var immediate: WidgetCollection? = null
	
	val synchronize: MutableSet<Identifier> = mutableSetOf()
	
	val hash: Int
		get() = Objects.hash(this, parent)
	
	open var hidden: Boolean = false
		get() = if (parent == null) field else field || parent!!.hidden
	
	open var focused: Boolean = false
		get() = if (parent == null) field && !hidden else field && !parent!!.hidden

	open var held: Boolean = false
		get() = if (parent == null) field else field && !parent!!.hidden

	open var parent: Widget? = null
	
	open fun onAdded(handled: WidgetCollection.Handled, immediate: WidgetCollection) {
		this.handled = handled
		this.immediate = immediate
	}

	open fun onRemoved(handled: WidgetCollection.Handled, immediate: WidgetCollection) {
		this.handled = null
		this.immediate = null
	}

	open fun onMouseMoved(x: Float, y: Float) {
		focus(x, y)

		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onMouseMoved(x, y)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.MouseMoved)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofMouseMove(handled!!.id!!, hash, x, y))
		}
	}

	open fun onMouseClicked(x: Float, y: Float, button: Int) {
		if (focused) held = true

		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onMouseClicked(x, y, button)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.MouseClicked)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofMouseClick(handled!!.id!!, hash, x, y, button))
		}
	}

	open fun onMouseReleased(x: Float, y: Float, button: Int) {
		held = false

		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onMouseReleased(x, y, button)
			}
		}
		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.MouseReleased)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofMouseRelease(handled!!.id!!, hash, x, y, button))
		}
	}

	open fun onMouseDragged(x: Float, y: Float, button: Int, deltaX: Double, deltaY: Double) {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onMouseDragged(x, y, button, deltaX, deltaY)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.MouseDragged)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofMouseDrag(handled!!.id!!, hash, x, y, button, deltaX, deltaY))
		}
	}

	open fun onMouseScrolled(x: Float, y: Float, deltaY: Double) {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onMouseScrolled(x, y, deltaY)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.MouseScrolled)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofMouseScroll(handled!!.id!!, hash, x, y, deltaY))
		}
	}

	open fun onKeyPressed(keyCode: Int, scanCode: Int, keyModifiers: Int) {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onKeyPressed(keyCode, scanCode, keyModifiers)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.KeyPressed)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofKeyPress(handled!!.id!!, hash, keyCode, scanCode, keyModifiers))
		}
	}

	open fun onKeyReleased(keyCode: Int, character: Int, keyModifiers: Int) {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onKeyReleased(keyCode, character, keyModifiers)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.KeyReleased)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofKeyRelease(handled!!.id!!, hash, keyCode, character, keyModifiers))
		}
	}

	open fun onCharacterTyped(character: Char, keyCode: Int) {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onCharacterTyped(character, keyCode)
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.CharacterTyped)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofCharType(handled!!.id!!, hash, character, keyCode))
		}
	}

	open fun onFocusGained() {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onFocusGained()
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.FocusGained)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofFocusGain(handled!!.id!!, hash))
		}
	}

	open fun onFocusReleased() {
		if (this is WidgetCollection && !(handled != null && !handled!!.client)) {
			widgets.forEach {
				it.onFocusReleased()
			}
		}

		if (focused && handled != null && handled!!.handler != null && handled!!.client && synchronize.contains(Networks.FocusReleased)) {
			Networks.toServer(Networks.WidgetUpdate, Networks.ofFocusRelease(handled!!.id!!, hash))
		}
	}

	@Environment(EnvType.CLIENT)
	open fun getTooltip(): List<Text> {
		return emptyList()
	}

	@Environment(EnvType.CLIENT)
	open fun drawWidget(matrices: MatrixStack, provider: VertexConsumerProvider) {
	
	}

	open fun onLayoutChanged() {
		focus(Positions.MouseX, Positions.MouseY)

		parent?.onLayoutChanged()
		handled?.onLayoutChanged()
	}

	fun focus(x: Float, y: Float) {
		val wasFocused = focused
		focused = isWithin(x, y)

		if (wasFocused && !focused) onFocusReleased()
		if (!wasFocused && focused) onFocusGained()
	}

	fun isWithin(x: Float, y: Float): Boolean {
		return x > position.x && x < position.x + size.width && y > position.y && y < position.y + size.height
	}
	
	open fun tick() {
		if (this is WidgetCollection) {
			widgets.forEach(Widget::tick)
		}
	}
	
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Widget) return false
		
		if (position != other.position) return false
		if (size != other.size) return false
		if (parent != other.parent) return false
		
		return true
	}
	
	override fun hashCode(): Int {
		var result = position.hashCode()
		result = 31 * result + size.hashCode()
		result = 31 * result + (parent?.hashCode() ?: 0)
		return result
	}
}